src/orion/storage/legacy.py

```diff
@@ -0,0 +1,119 @@
+    uri: str
+        database uri specifying how to connect to the database
+        the uri follows the following format
+        `mongodb://[username:password@]host1[:port1][,...hostN[:portN]][/[database][?options]]`
+
+    """
+
+    def __init__(self, uri=None):
+        self._db = Database()
+        self._setup_db()
+
+    def _setup_db(self):
+        """Database index setup"""
+        self._db.ensure_index('experiments',
+                              [('name', Database.ASCENDING),
+                               ('metadata.user', Database.ASCENDING)],
+                              unique=True)
+        self._db.ensure_index('experiments', 'metadata.datetime')
+
+        self._db.ensure_index('trials', 'experiment')
+        self._db.ensure_index('trials', 'status')
+        self._db.ensure_index('trials', 'results')
+        self._db.ensure_index('trials', 'start_time')
+        self._db.ensure_index('trials', [('end_time', Database.DESCENDING)])
+
+    def create_experiment(self, config):
+        """See :func:`~orion.storage.BaseStorageProtocol.create_experiment`"""
+        return self._db.write('experiments', config)
+
+    def update_experiment(self, experiment, where=None, **kwargs):
+        """See :func:`~orion.storage.BaseStorageProtocol.update_experiment`"""
+
+        if not kwargs:
+            return True
+
+        if where is None:
+            where = dict()
+
+        where['_id'] = experiment._id
+        return self._db.write('experiments', data=kwargs, query=where)
+
+    def fetch_experiments(self, query):
+        """See :func:`~orion.storage.BaseStorageProtocol.fetch_experiments`"""
+        return self._db.read('experiments', query)
+
+    def fetch_trials(self, query, selection=None):
+        """See :func:`~orion.storage.BaseStorageProtocol.fetch_trials`"""
+        return [Trial(**t) for t in self._db.read('trials', query=query, selection=selection)]
+
+    def register_trial(self, trial):
+        """See :func:`~orion.storage.BaseStorageProtocol.register_trial`"""
+        self._db.write('trials', trial.to_dict())
+        return trial
+
+    def retrieve_result(self, trial, results_file=None, **kwargs):
+        """Parse the result file that was generated by the trial process and
+        append the result to the trial object
+
+        Parameters
+        ----------
+        trial: Trial
+            The trial object to be updated
+
+        results_file: str
+            the file handle to read the result from
+
+        """
+        results = JSONConverter().parse(results_file.name)
+
+        trial.results = [
+            Trial.Result(
+                name=res['name'],
+                type=res['type'],
+                value=res['value']) for res in results
+        ]
+        return trial
+
+    def update_trial(self, trial: Trial, where=None, **kwargs) -> Trial:
+        """See :func:`~orion.storage.BaseStorageProtocol.update_trial`"""
+
+        if not kwargs:
+            return trial
```

Comment:
return False?

Reply:
Oh, sometimes it is None, and if write succeeds but read doesn't for some obscure reason then the output is the trial document from the write. :thinking: What about just updating and leaving it up to the external code to update the trial object if necessary?